### PR TITLE
chore: Create initial release process scripts for official ASF source release

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion-comet/issues/394

## Rationale for this change

We need to establish the basic ASF source release process

## What changes are included in this PR?

- Copy create-tarball, release-tarball, verify-release-candidate scripts from DataFusion and modify them for Comet
- Change the crate name from `comet` to `datafusion-comet` (`comet` was taken, and I have reserved `datafusion-comet` in crates.io)
- Add placeholder README

## How are these changes tested?

They are not tested yet. I will test them prior to our first release.

### CODE REVIEW

File: core/Cargo.toml
Comment (andygrove) [inline core/Cargo.toml:19]: There is already a `comet` crate, so I renamed this. I have reserved the crate: https://crates.io/crates/datafusion-comet

**Comment by andygrove** [comment]
@snmvaughan @parthchandra please review when you have the time

**Review by viirya — COMMENTED** [review]
Hmm, I think this is for publishing cargo package. But I think Comet should be released as a maven package at maven repository? So users can simply add Comet as a dependency in their pom file?

**Comment by andygrove** [comment]
> Hmm, I think this is for publishing cargo package. But I think Comet should be released as a maven package at maven repository? So users can simply add Comet as a dependency in their pom file?

It depends whether we want to allow other Rust projects to be able to use the Spark-compatible expressions that we are implementing, outside of the Comet/Spark integration.

The maven release will be needed for sure.

File: core/Cargo.toml
Comment (comphead) [inline core/Cargo.toml:24]:
```suggestion
description = "Apache DataFusion Comet: Apache Spark native query execution plugin"
```

This will be consistent with what in DataFusion known userlist

**Comment by andygrove** [comment]
> Hmm, I think this is for publishing cargo package. But I think Comet should be released as a maven package at maven repository? So users can simply add Comet as a dependency in their pom file?

Also, regardless of whether we choose to publish the crate or not, I think it is better to not use the name of an existing crate, and good to have the author information be consistent with other DataFusion crates.

**Comment by andygrove** [comment]
@viirya @comphead @parthchandra @kazuyukitanimura @huaxingao @advancedxy Could I get more feedback on this PR so that we can prepare for the upcoming 0.1.0 source release, which we are planning to do once we have switched back to depending on an official DataFusion release.

File: dev/release/verify-release-candidate.sh
Comment (andygrove) [reply dev/release/verify-release-candidate.sh:125]: Great point about testing the scala code. I will add that.

File: core/Cargo.toml
Comment (andygrove) [reply core/Cargo.toml:24]: It would be good to standardize our description

File: dev/release/check-rat-report.py
Comment (comphead) [inline dev/release/check-rat-report.py:11]: we do have a RAT check from maven, is this check a replacement or covers other usecase?

Comment (andygrove) [reply dev/release/check-rat-report.py:11]: It looks like we somehow missed adding the RAT check in maven. I was also surprised.

Comment (andygrove) [reply dev/release/check-rat-report.py:11]: @comphead I misread your comment. Where is the rat check in maven?

Comment (andygrove) [reply dev/release/check-rat-report.py:11]: nm, I see it now.

We need to run the rat check against the source tarball before we release it. The maven project is checking the contents of the repo not the release tarball, so maybe we need both. I will check how we are doing this in DataFusion.

Comment (comphead) [reply dev/release/check-rat-report.py:11]: yeah, ideally to remove duplications, perhaps we can remove RAT check in maven then and live only with the release one

File: dev/release/verify-release-candidate.sh
Comment (kazuyukitanimura) [inline dev/release/verify-release-candidate.sh:113]: Spark-3.4 is hardcoded now. What about making it a parameter.

Also does it make sense to use one of our make commands

**Review by kazuyukitanimura — COMMENTED** [review]
Mostly looks good. I feel we can merge and start using.

File: dev/release/check-rat-report.py
Comment (andygrove) [reply dev/release/check-rat-report.py:11]: DataFusion uses the same rat script from CI and from verify-release.sh. 

If we do the same with Comet and remove the maven task it could cause more work for developers because they would not catch issues as part of a standard maven build and would have to run a separate script. I am not sure if it is worth the effort to change this right now.

File: dev/release/verify-release-candidate.sh
Comment (andygrove) [reply dev/release/verify-release-candidate.sh:113]: We don't currently have a make command that specifies `-Dmaven.gitcommitid.skip=true` although one is being added in https://github.com/apache/datafusion-comet/pull/512

Comment (kazuyukitanimura) [reply dev/release/verify-release-candidate.sh:113]: Right, one hack could be something like `make release PROFILES="-Dmaven.gitcommitid.skip=true"`

File: core/Cargo.toml
Comment (parthchandra) [reply core/Cargo.toml:24]: I would vote for the description in the README. 
nit: high performance is two words (i.e. not hyphenated)

File: dev/release/create-tarball.sh
Comment (parthchandra) [inline dev/release/create-tarball.sh:132]: Can committers do this or only PMC?

Comment (parthchandra) [reply dev/release/create-tarball.sh:132]: We should probably add to the README that the entire release process requires PMC (or committer if that is sufficient).

File: dev/release/verify-release-candidate.sh
Comment (parthchandra) [reply dev/release/verify-release-candidate.sh:113]: Should we run tests as well? All comet tests pass for me when I run them though they do take a while to run.

Comment (parthchandra) [reply dev/release/verify-release-candidate.sh:113]: > Right, one hack could be something like `make release PROFILES="-Dmaven.gitcommitid.skip=true"`

That _would_ be a hack. I would leave it as it is. I assume that we will update/deprecate some of this once we move to publishing binary releases.

**Review by parthchandra — COMMENTED** [review]
LGTM. Some minor comments.

File: dev/release/create-tarball.sh
Comment (andygrove) [reply dev/release/create-tarball.sh:132]: Some steps can be completed by committers (such as creating the PR to update the project version number and generating the changelog) but some steps do require PMC members. 

I can add this to the README.

File: dev/release/verify-release-candidate.sh
Comment (andygrove) [reply dev/release/verify-release-candidate.sh:113]: > Should we run tests as well? All comet tests pass for me when I run them though they do take a while to run.

I think that the tests take too long. We create the release candidate tarball from a tag in GitHub where we have already run all of the tests, so running them all again is potentially redundant.

Perhaps there is some subset of tests that we could run? 🤔

File: dev/release/create-tarball.sh
Comment (andygrove) [reply dev/release/create-tarball.sh:132]: I have expanded the README to explain the different steps in the release process and who can do each part.

Comment (andygrove) [reply dev/release/create-tarball.sh:132]: I also added some better info on verifying release candidates.

File: dev/release/run-rat.sh
Comment (andygrove) [reply dev/release/run-rat.sh:35]: This shows the list of files that failed the rat test

File: dev/release/create-tarball.sh
Comment (andygrove) [reply dev/release/create-tarball.sh:132]: > I think committer permission should be sufficient? The bandwidth of PMC is limited, it would be great that we can make this committer sufficient. When releasing on apache uniffle, I don't think we need PMC to be the release manager.

Only PMC members have write access to the subversion repositories as far as I know.

Comment (parthchandra) [reply dev/release/create-tarball.sh:132]: Also, IIRC, when we do binary releases, for maven artifacts,  promoting release candidates to releases also requires PMC.

File: core/Cargo.toml
Comment (andygrove) [reply core/Cargo.toml]: I have updated this

File: dev/release/run-rat.sh
Comment (andygrove) [reply dev/release/run-rat.sh]: Good idea. I have made that change.

File: dev/release/verify-release-candidate.sh
Comment (andygrove) [reply dev/release/verify-release-candidate.sh]: > Spark-3.4 is hardcoded now. What about making it a parameter.

I added a comment explaining that we are testing the latest supported Spark version.

I think we can make more improvements in the future but I would like to get the basic scripts in place first so that we have the ability to make a release.

**Comment by parthchandra** [comment]
> @parthchandra @advancedxy I think I have addressed all of the feedback now

LGTM

**Comment by andygrove** [comment]
@viirya @kazuyukitanimura @huaxingao could I get a committer review/approval

File: dev/release/verify-release-candidate.sh
Comment (comphead) [inline dev/release/verify-release-candidate.sh:114]: does it mean if we support Spark 3.5, Spark4.0, we need to change this line? 
Wondering should we run it against all supported versions?

**Review by comphead — APPROVED** [review]
lgtm thanks @andygrove 
left a small comment

File: dev/release/verify-release-candidate.sh
Comment (andygrove) [reply dev/release/verify-release-candidate.sh:114]: It probably makes sense to verify against the latest Spark version here. I'm hoping that people verify release by building for their Spark version and running against their workflows. This script is just a basic verification that the release builds ok.

**Pull request merged**
